src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -354,10 +390,14 @@ public void replaceUnloadableService() throws ServiceLoadException {
         /**
          * Clean up services that the merge intends to remove.
          *
-         * @throws InterruptedException when the merge is interrupted
+         * @param totallyCompleteFuture deployment result future, used to check if a deployment is cancelled
+         * @throws InterruptedException when the merge is interrupted or cancelled
          * @throws ServiceUpdateException   when error is encountered while trying to close any service
+         * @throws DeploymentCancellationException if deployment is cancelled while closing services
          */
-        public void removeObsoleteServices() throws InterruptedException, ServiceUpdateException {
+        @SuppressWarnings("PMD.PreserveStackTrace")
+        public void removeObsoleteServices(CompletableFuture<DeploymentResult> totallyCompleteFuture)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 422-441, lines 442-454 into separate methods.